client/blocks/user-mentions/with-user-mentions.jsx

```diff
@@ -18,33 +22,250 @@ import React from 'react';
+				}
+
+				return this.hidePopover();
+			}
+
+			const query = this.getQueryText();
+
+			this.setState( {
+				showPopover: query !== null,
+				selectedSuggestionId: null,
+				query,
+			} );
+		};
+
+		getQueryText() {
+			const node = this.textInput.current;
+			// (?:^|\\s) means start of input or whitespace
+			// ([A-Za-z0-9_\+\-]*) means 0 or more instances of: A-Z a-z 0-9 _ + -
+			const matcher = new RegExp( '(?:^|\\s)@([A-Za-z0-9_+-]*)$', 'gi' );
+			const textBeforeCaret = node.value.slice( 0, node.selectionEnd );
+			const match = matcher.exec( textBeforeCaret );
+
+			return match && match.length > 1 ? match[ 1 ] : null;
+		}
+
+		getPosition() {
+			const node = this.textInput.current;
+			const nodeRect = node.getBoundingClientRect();
+			const caretPosition = getCaretCoordinates( node, node.selectionEnd );
+			const position = {
+				left: nodeRect.left + caretPosition.left,
+				top: nodeRect.top + caretPosition.top + 28,
+			};
+
+			console.log( position ); // eslint-disable-line no-console
+
+			return position;
+		}
+
+		getSuggestion() {
+			const index = this.getSelectedSuggestionIndex();
+
+			return index > -1 ? this.matchingSuggestions[ index ] : null;
+		}
+
+		getSelectedSuggestionIndex() {
+			if ( ! this.state.selectedSuggestionId ) {
+				return 0;
+			}
+
+			return findIndex(
+				this.matchingSuggestions,
+				( { ID: id } ) => id === this.state.selectedSuggestionId
+			);
+		}
+
+		getMatchingSuggestions( suggestions, query ) {
+			if ( query ) {
+				query = escapeRegExp( query );
+				const matcher = new RegExp( `^${ query }|\\s${ query }`, 'ig' ); // Start of string or preceded by a space.
+
+				suggestions = suggestions.filter( ( { user_login: login, display_name: name } ) =>
+					matcher.test( `${ login } ${ name }` )
+				);
+			}
+
+			return suggestions.slice( 0, 10 );
+		}
+
+		insertSuggestion = ( { user_login: userLogin } ) => {
+			if ( ! userLogin ) {
+				return;
+			}
+
+			const node = this.textInput.current;
+			const textBeforeCaret = node.value.slice( 0, node.selectionEnd );
+			const lastAtSymbolPosition = textBeforeCaret.lastIndexOf( '@' );
+			const textBeforeAtSymbol = node.value.slice( 0, lastAtSymbolPosition );
+			const textAfterSelectionEnd = node.value.slice( node.selectionEnd, node.value.length + 1 );
+
+			node.value = textBeforeAtSymbol + '@' + userLogin + textAfterSelectionEnd;
+		};
+
+		updatePosition = ( state = this.state, newPosition ) => {
+			if ( ! newPosition ) {
+				newPosition = this.getPosition( state );
+			}
+
+			this.setState( { popoverPosition: newPosition } );
+		};
+
+		throttledUpdatePosition = throttle( this.updatePosition, 100 );
+
+		hidePopover = () => this.setState( { showPopover: false } );
+
 		render() {
-			// const suggestions = [
-			// 	{
-			// 		ID: 1,
-			// 		user_login: 'testuser',
-			// 	},
-			// ];
+			//const { siteId } = this.props;
+			const { query, showPopover } = this.state;
+			const suggestions = [
```

Comment:
It might be useful to specify this using `defaultProp.suggestions`. That way, it'll be easy to swap out the placeholder suggestions in the future.

Reply:
That's a smart idea :)